sdk-rs/src/lib.rs

```diff
@@ -0,0 +1,790 @@
+fn build_accounts(
+    context: Context,
+    base_accounts: impl ToAccountMetas,
+    user: &User,
+    markets_readable: &[MarketId],
+    markets_writable: &[MarketId],
+) -> Vec<AccountMeta> {
+    // the order of accounts returned must be instruction, oracles, spot, perps see (https://github.com/drift-labs/protocol-v2/blob/master/programs/drift/src/instructions/optional_accounts.rs#L28)
+    let mut seen = [0_u64; 2]; // [spot, perp]
+    let mut accounts = Vec::<AccountType>::default();
+
+    // add accounts to the ordered list
+    let mut include_market = |market_index: u16, market_type: MarketType, writable: bool| {
+        let index_bit = 1_u64 << market_index as u8;
+        // always safe since market type is 0 or 1
+        let seen_by_type = unsafe { seen.get_unchecked_mut(market_type as usize % 2) };
+        if *seen_by_type & index_bit > 0 {
+            return;
+        }
+        *seen_by_type |= index_bit;
+
+        let (account, oracle) = match market_type {
+            MarketType::Spot => {
+                let SpotMarketConfig {
+                    account, oracle, ..
+                } = constants::spot_market_by_index(context, market_index);
+                (
+                    AccountType::Spot {
+                        pubkey: *account,
+                        writable,
+                    },
+                    oracle,
+                )
+            }
+            MarketType::Perp => {
+                let PerpMarketConfig {
+                    account, oracle, ..
+                } = constants::perp_market_by_index(context, market_index);
+                (
+                    AccountType::Perp {
+                        pubkey: *account,
+                        writable,
+                    },
+                    oracle,
+                )
+            }
+        };
+        if let Err(idx) = accounts.binary_search(&account) {
+            accounts.insert(idx, account);
+        }
+        let oracle = AccountType::Oracle { pubkey: *oracle };
+        if let Err(idx) = accounts.binary_search(&oracle) {
+            accounts.insert(idx, oracle);
+        }
+    };
+
+    for MarketId { index, kind } in markets_writable {
+        include_market(*index, *kind, true);
+    }
+
+    for MarketId { index, kind } in markets_readable {
+        include_market(*index, *kind, false);
+    }
+
+    // Drift program performs margin checks which requires reading user positions
+    for p in user.spot_positions.iter().filter(|p| !p.is_available()) {
+        include_market(p.market_index, MarketType::Spot, false);
+    }
+    for p in user.perp_positions.iter().filter(|p| !p.is_available()) {
+        include_market(p.market_index, MarketType::Perp, false);
+    }
+
+    // always manually try to include the quote (USDC) market
+    // TODO: this is not exactly the same semantics as the TS sdk
+    include_market(QUOTE_SPOT_MARKET_INDEX, MarketType::Spot, false);
```

Comment:
yeah if you just loaded the quote_asset_market-index on the perp market account, you wouldn't need this